core/corehttp/readers.go

```diff
@@ -0,0 +1,134 @@
+package corehttp
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/ipfs/go-ipfs/commands/files"
+	"github.com/ipfs/go-ipfs/importer"
+	chunk "github.com/ipfs/go-ipfs/importer/chunk"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+)
+
+// helper functions for the gateway handler
+
+func (i *gatewayHandler) newDagFromReader(r io.Reader) (*dag.Node, error) {
+	// TODO(cryptix): change and remove this helper once PR1136 is merged
+	// return ufs.AddFromReader(i.node, r.Body)
+	return importer.BuildDagFromReader(
+		r, i.node.DAG, chunk.DefaultSplitter, importer.BasicPinnerCB(i.node.Pinning.GetManual()))
+}
+
+// add file, or directory recursively
+func (i *gatewayHandler) addFileRecursive(file files.File) (*dag.Node, error) {
+	if file.IsDirectory() {
+		return i.addDirRecursive(file)
+	}
+	return i.newDagFromReader(file)
+}
+
+// add a directory recursively
+func (i *gatewayHandler) addDirRecursive(file files.File) (*dag.Node, error) {
```

Comment:
https://github.com/ipfs/go-ipfs/blob/master/core/commands/add.go#L329
...if common functions in corehttp and commands can be deduplicated?